### PR TITLE
Don't use a generator for BoxedResolver 

### PR DESCRIPTION
**Pull request by bjorn3**
The generator is non-trivial and requires unsafe code anyway. Using regular unsafe code without a generator is much easier to follow.

Based on #85810 as it touches rustc_interface too.

### CODE REVIEW

File: compiler/rustc_interface/src/passes.rs [hunk 1]
```diff
@@ -85,11 +87,74 @@ fn count_nodes(krate: &ast::Crate) -> usize {
...
+                &'a ResolverArenas<'a>,
+            ) -> Result<(ast::Crate, Resolver<'a>)>,
+        {
+            let mut boxed_resolver = Box::new(BoxedResolverInner {
+                session,
+                resolver_arenas: Resolver::arenas(),
+                resolver: None,
+                _pin: PhantomPinned,
+            });
+            unsafe {
```
Comment (cjgillot) [inline compiler/rustc_interface/src/passes.rs:121]: Should there be a `SAFETY` comment here?

File: compiler/rustc_interface/src/passes.rs [hunk 2]
```diff
@@ -85,11 +87,74 @@ fn count_nodes(krate: &ast::Crate) -> usize {
...
+                    ),
+                )?;
+                boxed_resolver.resolver =
+                    Some(std::mem::transmute::<Resolver<'_>, Resolver<'_>>(resolver));
+                Ok((crate_, BoxedResolver(Pin::new_unchecked(boxed_resolver))))
+            }
+        }
+
+        pub fn access<F: for<'a> FnOnce(&mut Resolver<'a>) -> R, R>(&mut self, f: F) -> R {
+            let mut resolver = unsafe {
```
Comment (cjgillot) [inline compiler/rustc_interface/src/passes.rs:135]: Should there be a `SAFETY` comment here?

File: compiler/rustc_interface/src/passes.rs [hunk 3]
```diff
@@ -85,11 +87,74 @@ fn count_nodes(krate: &ast::Crate) -> usize {
...
+                self.0.as_mut().map_unchecked_mut(|boxed_resolver| &mut boxed_resolver.resolver)
+            };
+            f((&mut *resolver).as_mut().unwrap())
+        }
+
+        pub fn to_resolver_outputs(resolver: Rc<RefCell<BoxedResolver>>) -> ResolverOutputs {
+            match Rc::try_unwrap(resolver) {
+                Ok(resolver) => {
+                    let mut resolver = resolver.into_inner();
+                    let mut resolver = unsafe {
```
Comment (cjgillot) [inline compiler/rustc_interface/src/passes.rs:145]: Should there be a `SAFETY` comment here?

**Review by cjgillot — APPROVED** [review]
Overall LGTM, and a nice simplification.

File: compiler/rustc_interface/src/passes.rs [hunk 4]
```diff
@@ -85,11 +87,74 @@ fn count_nodes(krate: &ast::Crate) -> usize {
...
+            });
+            unsafe {
+                let (crate_, resolver) = make_resolver(
+                    std::mem::transmute::<&Session, &Session>(&boxed_resolver.session),
+                    std::mem::transmute::<&ResolverArenas<'_>, &ResolverArenas<'_>>(
+                        &boxed_resolver.resolver_arenas,
+                    ),
+                )?;
+                boxed_resolver.resolver =
+                    Some(std::mem::transmute::<Resolver<'_>, Resolver<'_>>(resolver));
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply compiler/rustc_interface/src/passes.rs:129]: Yes, it casts from the `'a` representing the arena lifetime to a dummy `'static`.

[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_interface/src/passes.rs:129]: Isn't the arena lifetime set to `'static` by a previous transmute?

[the remark this answers is not in the document]
Comment (bjorn3) [reply compiler/rustc_interface/src/passes.rs:129]: No, `make_resolver` takes an area transmuted from `'static` to `'a` as arg and returns a resolver with `'a` as lifetime.

File: compiler/rustc_interface/src/passes.rs [hunk 5]
```diff
@@ -85,11 +87,74 @@ fn count_nodes(krate: &ast::Crate) -> usize {
...
+                Ok((crate_, BoxedResolver(Pin::new_unchecked(boxed_resolver))))
+            }
+        }
+
+        pub fn access<F: for<'a> FnOnce(&mut Resolver<'a>) -> R, R>(&mut self, f: F) -> R {
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply compiler/rustc_interface/src/passes.rs:134]: Not sure, most self-referential struct crates use a closure.

File: compiler/rustc_interface/src/passes.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_interface/src/passes.rs:137]: The transmute from the local lifetime `'boxed_resolver` to `'static` is already performed by the two transmutes above. The call to make_resolver substitutes quantified `'a` into `'static`. Therefore, this last one should not be necessary.

[the remark this answers is not in the document]
Comment (bjorn3) [reply compiler/rustc_interface/src/passes.rs:137]: Ah, I see.

[the remark this answers is not in the document]
Comment (bjorn3) [reply compiler/rustc_interface/src/passes.rs]: Removed

**Pull request merged**